lib/puppet/ssl/state_machine.rb

```diff
@@ -84,6 +98,51 @@ def next_state
         to_error(_('Could not download CA certificate: %{message}') % { message: e.message }, e)
       end
     end
+
+    private
+
+    def needs_refresh?(now, last_update)
+      return true if last_update.nil?
+
+      ca_ttl = Puppet[:ca_refresh_interval]
+      return false unless ca_ttl
+
+      now.to_i > last_update.to_i + ca_ttl
+    end
+
+    def refresh_ca(ssl_ctx, last_update)
+      Puppet.info(_("Refreshing CA certificate"))
+
+      # return the next_ctx containing the updated ca
+      [download_ca(ssl_ctx, last_update), true]
+    rescue Puppet::HTTP::ResponseError => e
```

Comment:
Feels outside of the context of this PR, but it seems strange to me that we would raise a 304 response as a `Puppet::HTTP::ResponseError`.

Reply:
Some of puppet's http client was inspired by httpclient, include how to raises for responses other than 2xx: https://github.com/nahi/httpclient/blob/4658227a46f7caa633ef8036f073bbd1f0a955a2/lib/httpclient.rb#L1127-L1133